File: source/math/mat4.cpp
#include "math/mat4.h"

#include <iostream>

#include "utility/convert.h"


kl::mat4::mat4()
{
	setIdentity();
}

// Getter
float& kl::mat4::operator[](int i)
{
	return data[i];
}

// Addition
kl::mat4 kl::mat4::add(const kl::mat4& obj) const
{
	kl::mat4 temp;
	for (int i = 0; i < 16; i++)
		temp[i] = data[i] + obj.data[i];
	return temp;
}
kl::mat4 kl::mat4::operator+(const kl::mat4& obj) const
{
	return this->add(obj);
}
void kl::mat4::operator+=(const kl::mat4& obj)
{
	for (int i = 0; i < 16; i++)
		data[i] += obj.data[i];
}

// Subtraction
kl::mat4 kl::mat4::sub(const kl::mat4& obj) const
{
	kl::mat4 temp;
	for (int i = 0; i < 16; i++)
		temp[i] = data[i] - obj.data[i];
	return temp;
}
kl::mat4 kl::mat4::operator-(const kl::mat4& obj) const
{
	return this->sub(obj);
}
void kl::mat4::operator-=(const kl::mat4& obj)
{
	for (int i = 0; i < 16; i++)
		data[i] -= obj.data[i];
}

// Multiplication
kl::mat4 kl::mat4::mul(float a) const
{
	kl::mat4 temp;
	for (int i = 0; i < 16; i++)
		temp[i] = data[i] * a;
	return temp;
}
kl::mat4 kl::mat4::operator*(float a) const
{
	return this->mul(a);
}
void kl::mat4::operator*=(float a)
{
	for (int i = 0; i < 16; i++)
		data[i] *= a;
}
kl::mat4 kl::mat4::mul(const kl::mat4& obj) const
{
	kl::mat4 temp;
	temp[0] = data[0] * obj.data[0] + data[1] * obj.data[4] + data[2] * obj.data[8] + data[3] * obj.data[12];
	temp[1] = data[0] * obj.data[1] + data[1] * obj.data[5] + data[2] * obj.data[9] + data[3] * obj.data[13];
	temp[2] = data[0] * obj.data[2] + data[1] * obj.data[6] + data[2] * obj.data[10] + data[3] * obj.data[14];
	temp[3] = data[0] * obj.data[3] + data[1] * obj.data[7] + data[2] * obj.data[11] + data[3] * obj.data[15];
	temp[4] = data[4] * obj.data[0] + data[5] * obj.data[4] + data[6] * obj.data[8] + data[7] * obj.data[12];
	temp[5] = data[4] * obj.data[1] + data[5] * obj.data[5] + data[6] * obj.data[9] + data[7] * obj.data[13];
	temp[6] = data[4] * obj.data[2] + data[5] * obj.data[6] + data[6] * obj.data[10] + data[7] * obj.data[14];
	temp[7] = data[4] * obj.data[3] + data[5] * obj.data[7] + data[6] * obj.data[11] + data[7] * obj.data[15];
	temp[8] = data[8] * obj.data[0] + data[9] * obj.data[4] + data[10] * obj.data[8] + data[11] * obj.data[12];
	temp[9] = data[8] * obj.data[1] + data[9] * obj.data[5] + data[10] * obj.data[9] + data[11] * obj.data[13];
	temp[10] = data[8] * obj.data[2] + data[9] * obj.data[6] + data[10] * obj.data[10] + data[11] * obj.data[14];
	temp[11] = data[8] * obj.data[3] + data[9] * obj.data[7] + data[10] * obj.data[11] + data[11] * obj.data[15];
	temp[12] = data[12] * obj.data[0] + data[13] * obj.data[4] + data[14] * obj.data[8] + data[15] * obj.data[12];
	temp[13] = data[12] * obj.data[1] + data[13] * obj.data[5] + data[14] * obj.data[9] + data[15] * obj.data[13];
	temp[14] = data[12] * obj.data[2] + data[13] * obj.data[6] + data[14] * obj.data[10] + data[15] * obj.data[14];
	temp[15] = data[12] * obj.data[3] + data[13] * obj.data[7] + data[14] * obj.data[11] + data[15] * obj.data[15];
	return temp;
}
kl::mat4 kl::mat4::operator*(const kl::mat4& obj) const
{
	return this->mul(obj);
}
void kl::mat4::operator*=(const kl::mat4& obj)
{
	*this = this->mul(obj);
}
kl::float4 kl::mat4::mul(const kl::float4& obj) const
{
	kl::float4 temp;
	temp.x = data[0] * obj.x + data[1] * obj.y + data[2] * obj.z + data[3] * obj.w;
	temp.y = data[4] * obj.x + data[5] * obj.y + data[6] * obj.z + data[7] * obj.w;
	temp.z = data[8] * obj.x + data[9] * obj.y + data[10] * obj.z + data[11] * obj.w;
	temp.w = data[12] * obj.x + data[13] * obj.y + data[14] * obj.z + data[15] * obj.w;
	return temp;
}
kl::float4 kl::mat4::operator*(const kl::float4& obj) const
{
	return this->mul(obj);
}

// Division
kl::mat4 kl::mat4::div(float a) const
{
	return this->mul(1 / a);
}
kl::mat4 kl::mat4::operator/(float a) const
{
	return this->div(a);
}
void kl::mat4::operator /= (float a)
{
	this->operator*=(1 / a);
}

// Comparison
bool kl::mat4::equals(const kl::mat4& obj) const
{
	for (int i = 0; i < 16; i++)
	{
		if (data[i] != obj.data[i])
			return false;
	}
	return true;
}
bool kl::mat4::operator==(const kl::mat4& obj) const
{
	return this->equals(obj);
}
bool kl::mat4::operator!=(const kl::mat4& obj) const
{
	return !this->equals(obj);
}

// Returns pointer to raw data
float* kl::mat4::pointer() const
{
	return (float*)data;
}

// Loads the identity matrix
void kl::mat4::setIdentity()
{
	data[0] = 1; data[1] = 0; data[2] = 0; data[3] = 0;
	data[4] = 0; data[5] = 1; data[6] = 0; data[7] = 0;
	data[8] = 0; data[9] = 0; data[10] = 1; data[11] = 0;
	data[12] = 0; data[13] = 0; data[14] = 0; data[15] = 1;
}

// Returns inverse matrix if possible
kl::mat4 kl::mat4::inverse() const
{
	kl::mat4 inv;
	inv[0] = data[5] * data[10] * data[15] -
		data[5] * data[11] * data[14] -
		data[9] * data[6] * data[15] +
		data[9] * data[7] * data[14] +
		data[13] * data[6] * data[11] -
		data[13] * data[7] * data[10];
	inv[4] = -data[4] * data[10] * data[15] +
		data[4] * data[11] * data[14] +
		data[8] * data[6] * data[15] -
		data[8] * data[7] * data[14] -
		data[12] * data[6] * data[11] +
		data[12] * data[7] * data[10];
	inv[8] = data[4] * data[9] * data[15] -
		data[4] * data[11] * data[13] -
		data[8] * data[5] * data[15] +
		data[8] * data[7] * data[13] +
		data[12] * data[5] * data[11] -
		data[12] * data[7] * data[9];
	inv[12] = -data[4] * data[9] * data[14] +
		data[4] * data[10] * data[13] +
		data[8] * data[5] * data[14] -
		data[8] * data[6] * data[13] -
		data[12] * data[5] * data[10] +
		data[12] * data[6] * data[9];
	inv[1] = -data[1] * data[10] * data[15] +
		data[1] * data[11] * data[14] +
		data[9] * data[2] * data[15] -
		data[9] * data[3] * data[14] -
		data[13] * data[2] * data[11] +
		data[13] * data[3] * data[10];
	inv[5] = data[0] * data[10] * data[15] -
		data[0] * data[11] * data[14] -
		data[8] * data[2] * data[15] +
		data[8] * data[3] * data[14] +
		data[12] * data[2] * data[11] -
		data[12] * data[3] * data[10];
	inv[9] = -data[0] * data[9] * data[15] +
		data[0] * data[11] * data[13] +
		data[8] * data[1] * data[15] -
		data[8] * data[3] * data[13] -
		data[12] * data[1] * data[11] +
		data[12] * data[3] * data[9];
	inv[13] = data[0] * data[9] * data[14] -
		data[0] * data[10] * data[13] -
		data[8] * data[1] * data[14] +
		data[8] * data[2] * data[13] +
		data[12] * data[1] * data[10] -
		data[12] * data[2] * data[9];
	inv[2] = data[1] * data[6] * data[15] -
		data[1] * data[7] * data[14] -
		data[5] * data[2] * data[15] +
		data[5] * data[3] * data[14] +
		data[13] * data[2] * data[7] -
		data[13] * data[3] * data[6];
	inv[6] = -data[0] * data[6] * data[15] +
		data[0] * data[7] * data[14] +
		data[4] * data[2] * data[15] -
		data[4] * data[3] * data[14] -
		data[12] * data[2] * data[7] +
		data[12] * data[3] * data[6];
	inv[10] = data[0] * data[5] * data[15] -
		data[0] * data[7] * data[13] -
		data[4] * data[1] * data[15] +
		data[4] * data[3] * data[13] +
		data[12] * data[1] * data[7] -
		data[12] * data[3] * data[5];
	inv[14] = -data[0] * data[5] * data[14] +
		data[0] * data[6] * data[13] +
		data[4] * data[1] * data[14] -
		data[4] * data[2] * data[13] -
		data[12] * data[1] * data[6] +
		data[12] * data[2] * data[5];
	inv[3] = -data[1] * data[6] * data[11] +
		data[1] * data[7] * data[10] +
		data[5] * data[2] * data[11] -
		data[5] * data[3] * data[10] -
		data[9] * data[2] * data[7] +
		data[9] * data[3] * data[6];
	inv[7] = data[0] * data[6] * data[11] -
		data[0] * data[7] * data[10] -
		data[4] * data[2] * data[11] +
		data[4] * data[3] * data[10] +
		data[8] * data[2] * data[7] -
		data[8] * data[3] * data[6];
	inv[11] = -data[0] * data[5] * data[11] +
		data[0] * data[7] * data[9] +
		data[4] * data[1] * data[11] -
		data[4] * data[3] * data[9] -
		data[8] * data[1] * data[7] +
		data[8] * data[3] * data[5];
	inv[15] = data[0] * data[5] * data[10] -
		data[0] * data[6] * data[9] -
		data[4] * data[1] * data[10] +
		data[4] * data[2] * data[9] +
		data[8] * data[1] * data[6] -
		data[8] * data[2] * data[5];

	// Det calc
	const float det = data[0] * inv[0] + data[1] * inv[4] + data[2] * inv[8] + data[3] * inv[12];

	// Det check
	if (det == 0.0f)
		return kl::mat4();

	// Return
	return inv.div(det);
}

// Overloading std::cout
std::ostream& kl::operator<<(std::ostream& os, kl::mat4& obj)
{
	os <<
		kl::convert::toString(obj[0]) << " " << kl::convert::toString(obj[1]) << " " << kl::convert::toString(obj[2]) << " " << kl::convert::toString(obj[3]) << "\n" <<
		kl::convert::toString(obj[4]) << " " << kl::convert::toString(obj[5]) << " " << kl::convert::toString(obj[6]) << " " << kl::convert::toString(obj[7]) << "\n" <<
		kl::convert::toString(obj[8]) << " " << kl::convert::toString(obj[9]) << " " << kl::convert::toString(obj[10]) << " " << kl::convert::toString(obj[11]) << "\n" <<
		kl::convert::toString(obj[12]) << " " << kl::convert::toString(obj[13]) << " " << kl::convert::toString(obj[14]) << " " << kl::convert::toString(obj[15]);
	return os;
}
std::ostream& kl::operator<<(std::ostream& os, kl::mat4&& obj)
{
	os <<
		kl::convert::toString(obj[0]) << " " << kl::convert::toString(obj[1]) << " " << kl::convert::toString(obj[2]) << " " << kl::convert::toString(obj[3]) << "\n" <<
		kl::convert::toString(obj[4]) << " " << kl::convert::toString(obj[5]) << " " << kl::convert::toString(obj[6]) << " " << kl::convert::toString(obj[7]) << "\n" <<
		kl::convert::toString(obj[8]) << " " << kl::convert::toString(obj[9]) << " " << kl::convert::toString(obj[10]) << " " << kl::convert::toString(obj[11]) << "\n" <<
		kl::convert::toString(obj[12]) << " " << kl::convert::toString(obj[13]) << " " << kl::convert::toString(obj[14]) << " " << kl::convert::toString(obj[15]);
	return os;
}

// Returns a translation matrix
kl::mat4 kl::mat4::translate(const kl::float3& translation)
{
	kl::mat4 temp;
	temp[3] = translation.x;
	temp[7] = translation.y;
	temp[11] = translation.z;
	return temp;
}

// Returns a rotation matrix
kl::mat4 kl::mat4::rotate(const kl::float3& rotation)
{
	// Computing x rotation matrix
	const float xRad = kl::convert::toRadians(rotation.x);
	const float xSin = sin(xRad);
	const float xCos = cos(xRad);
	kl::mat4 xRot;
	xRot[5] = xCos;
	xRot[6] = -xSin;
	xRot[9] = xSin;
	xRot[10] = xCos;

	// Computing y rotation matrix
	const float yRad = kl::convert::toRadians(rotation.y);
	const float ySin = sin(yRad);
	const float yCos = cos(yRad);
	kl::mat4 yRot;
	yRot[0] = yCos;
	yRot[2] = ySin;
	yRot[8] = -ySin;
	yRot[10] = yCos;

	// Computing z rotation matrix
	const float zRad = kl::convert::toRadians(rotation.z);
	const float zSin = sin(zRad);
	const float zCos = cos(zRad);
	kl::mat4 zRot;
	zRot[0] = zCos;
	zRot[1] = -zSin;
	zRot[4] = zSin;
	zRot[5] = zCos;

	// Multiplying and returning
	return zRot * yRot * xRot;
}

// Returns a scaling matrix
kl::mat4 kl::mat4::scale(const kl::float3& size)
{
	kl::mat4 temp;
	temp[0] = size.x;
	temp[5] = size.y;
	temp[10] = size.z;
	return temp;
}

// Returns the perspective projection matrix
kl::mat4 kl::mat4::persp(float fov, float ar, float zNear, float zFar)
{
	const float tanHalf = 1 / tan(kl::convert::toRadians(fov) * 0.5f);
	kl::mat4 temp;
	temp[0] = tanHalf / ar;
	temp[5] = tanHalf;
	temp[10] = (-zFar - zNear) / (zNear - zFar);
	temp[11] = (2 * zNear * zFar) / (zNear - zFar);
	temp[14] = 1;
	temp[15] = 0;
	return temp;
}

// Returns the orthographics projection matrix
kl::mat4 kl::mat4::ortho(float left, float right, float bottom, float top, float nearZ, float farZ)
{
	kl::mat4 temp;
	temp[0] = 2 / (right - left);
	temp[5] = 2 / (top - bottom);
	temp[10] = -2 / (farZ - nearZ);
	temp[3] = -(right + left) / (right - left);
	temp[7] = -(top + bottom) / (top - bottom);
	temp[11] = -(farZ + nearZ) / (farZ - nearZ);
	return temp;
}

// Returns the "look at" matrix
kl::mat4 kl::mat4::lookAt(const kl::float3& position, const kl::float3& target, const kl::float3& up)
{
	// Vector calculations
	const kl::float3 f = (target - position).normalize();
	const kl::float3 s = up.cross(f).normalize();
	const kl::float3 u = f.cross(s);

	// Matrix building
	kl::mat4 temp;
	temp[0] = s.x;
	temp[1] = s.y;
	temp[2] = s.z;
	temp[3] = -s.dot(position);
	temp[4] = u.x;
	temp[5] = u.y;
	temp[6] = u.z;
	temp[7] = -u.dot(position);
	temp[8] = f.x;
	temp[9] = f.y;
	temp[10] = f.z;
	temp[11] = -f.dot(position);
	return temp;
}

File: source/kl/math/Mat4.java
package kl.math;

public class Mat4 {
    public final float[] data = new float[16];

    public Mat4() {
        data[ 0] = 1.0f; data[ 1] = 0.0f; data[ 2] = 0.0f; data[ 3] = 0.0f;
        data[ 4] = 0.0f; data[ 5] = 1.0f; data[ 6] = 0.0f; data[ 7] = 0.0f;
        data[ 8] = 0.0f; data[ 9] = 0.0f; data[10] = 1.0f; data[11] = 0.0f;
        data[12] = 0.0f; data[13] = 0.0f; data[14] = 0.0f; data[15] = 1.0f;
    }
    public Mat4(Mat4 m) {
        System.arraycopy(m.data, 0, data, 0, data.length);
    }

    public Mat4 add(Mat4 mat) {
        Mat4 temp = new Mat4();
        for (int i = 0; i < 16; i++) {
            temp.data[i] = data[i] + mat.data[i];
        }
        return temp;
    }

    public Mat4 subtract(Mat4 mat) {
        Mat4 temp = new Mat4();
        for (int i = 0; i < 16; i++) {
            temp.data[i] = data[i] - mat.data[i];
        }
        return temp;
    }

    public Mat4 multiply(float a) {
        Mat4 temp = new Mat4();
        for (int i = 0; i < 16; i++) {
            temp.data[i] = data[i] * a;
        }
        return temp;
    }
    public Float4 multiply(Float4 vec) {
        Float4 temp = new Float4(0);
        temp.x = data[ 0] * vec.x + data[ 1] * vec.y + data[ 2] * vec.z + data[ 3] * vec.w;
        temp.y = data[ 4] * vec.x + data[ 5] * vec.y + data[ 6] * vec.z + data[ 7] * vec.w;
        temp.z = data[ 8] * vec.x + data[ 9] * vec.y + data[10] * vec.z + data[11] * vec.w;
        temp.w = data[12] * vec.x + data[13] * vec.y + data[14] * vec.z + data[15] * vec.w;
        return temp;
    }
    public Mat4 multiply(Mat4 mat) {
        Mat4 temp = new Mat4();
        temp.data[ 0] = data[ 0] * mat.data[ 0] + data[ 1] * mat.data[ 4] + data[ 2] * mat.data[ 8] + data[ 3] * mat.data[12];
        temp.data[ 1] = data[ 0] * mat.data[ 1] + data[ 1] * mat.data[ 5] + data[ 2] * mat.data[ 9] + data[ 3] * mat.data[13];
        temp.data[ 2] = data[ 0] * mat.data[ 2] + data[ 1] * mat.data[ 6] + data[ 2] * mat.data[10] + data[ 3] * mat.data[14];
        temp.data[ 3] = data[ 0] * mat.data[ 3] + data[ 1] * mat.data[ 7] + data[ 2] * mat.data[11] + data[ 3] * mat.data[15];
        temp.data[ 4] = data[ 4] * mat.data[ 0] + data[ 5] * mat.data[ 4] + data[ 6] * mat.data[ 8] + data[ 7] * mat.data[12];
        temp.data[ 5] = data[ 4] * mat.data[ 1] + data[ 5] * mat.data[ 5] + data[ 6] * mat.data[ 9] + data[ 7] * mat.data[13];
        temp.data[ 6] = data[ 4] * mat.data[ 2] + data[ 5] * mat.data[ 6] + data[ 6] * mat.data[10] + data[ 7] * mat.data[14];
        temp.data[ 7] = data[ 4] * mat.data[ 3] + data[ 5] * mat.data[ 7] + data[ 6] * mat.data[11] + data[ 7] * mat.data[15];
        temp.data[ 8] = data[ 8] * mat.data[ 0] + data[ 9] * mat.data[ 4] + data[10] * mat.data[ 8] + data[11] * mat.data[12];
        temp.data[ 9] = data[ 8] * mat.data[ 1] + data[ 9] * mat.data[ 5] + data[10] * mat.data[ 9] + data[11] * mat.data[13];
        temp.data[10] = data[ 8] * mat.data[ 2] + data[ 9] * mat.data[ 6] + data[10] * mat.data[10] + data[11] * mat.data[14];
        temp.data[11] = data[ 8] * mat.data[ 3] + data[ 9] * mat.data[ 7] + data[10] * mat.data[11] + data[11] * mat.data[15];
        temp.data[12] = data[12] * mat.data[ 0] + data[13] * mat.data[ 4] + data[14] * mat.data[ 8] + data[15] * mat.data[12];
        temp.data[13] = data[12] * mat.data[ 1] + data[13] * mat.data[ 5] + data[14] * mat.data[ 9] + data[15] * mat.data[13];
        temp.data[14] = data[12] * mat.data[ 2] + data[13] * mat.data[ 6] + data[14] * mat.data[10] + data[15] * mat.data[14];
        temp.data[15] = data[12] * mat.data[ 3] + data[13] * mat.data[ 7] + data[14] * mat.data[11] + data[15] * mat.data[15];
        return temp;
    }

    public Mat4 div(float a) {
        return multiply(1.0f / a);
    }

    public Mat4 inverse() {
        Mat4 temp = new Mat4();
        temp.data[0] = data[5] * data[10] * data[15] -
                data[5] * data[11] * data[14] -
                data[9] * data[6] * data[15] +
                data[9] * data[7] * data[14] +
                data[13] * data[6] * data[11] -
                data[13] * data[7] * data[10];
        temp.data[4] = -data[4] * data[10] * data[15] +
                data[4] * data[11] * data[14] +
                data[8] * data[6] * data[15] -
                data[8] * data[7] * data[14] -
                data[12] * data[6] * data[11] +
                data[12] * data[7] * data[10];
        temp.data[8] = data[4] * data[9] * data[15] -
                data[4] * data[11] * data[13] -
                data[8] * data[5] * data[15] +
                data[8] * data[7] * data[13] +
                data[12] * data[5] * data[11] -
                data[12] * data[7] * data[9];
        temp.data[12] = -data[4] * data[9] * data[14] +
                data[4] * data[10] * data[13] +
                data[8] * data[5] * data[14] -
                data[8] * data[6] * data[13] -
                data[12] * data[5] * data[10] +
                data[12] * data[6] * data[9];
        temp.data[1] = -data[1] * data[10] * data[15] +
                data[1] * data[11] * data[14] +
                data[9] * data[2] * data[15] -
                data[9] * data[3] * data[14] -
                data[13] * data[2] * data[11] +
                data[13] * data[3] * data[10];
        temp.data[5] = data[0] * data[10] * data[15] -
                data[0] * data[11] * data[14] -
                data[8] * data[2] * data[15] +
                data[8] * data[3] * data[14] +
                data[12] * data[2] * data[11] -
                data[12] * data[3] * data[10];
        temp.data[9] = -data[0] * data[9] * data[15] +
                data[0] * data[11] * data[13] +
                data[8] * data[1] * data[15] -
                data[8] * data[3] * data[13] -
                data[12] * data[1] * data[11] +
                data[12] * data[3] * data[9];
        temp.data[13] = data[0] * data[9] * data[14] -
                data[0] * data[10] * data[13] -
                data[8] * data[1] * data[14] +
                data[8] * data[2] * data[13] +
                data[12] * data[1] * data[10] -
                data[12] * data[2] * data[9];
        temp.data[2] = data[1] * data[6] * data[15] -
                data[1] * data[7] * data[14] -
                data[5] * data[2] * data[15] +
                data[5] * data[3] * data[14] +
                data[13] * data[2] * data[7] -
                data[13] * data[3] * data[6];
        temp.data[6] = -data[0] * data[6] * data[15] +
                data[0] * data[7] * data[14] +
                data[4] * data[2] * data[15] -
                data[4] * data[3] * data[14] -
                data[12] * data[2] * data[7] +
                data[12] * data[3] * data[6];
        temp.data[10] = data[0] * data[5] * data[15] -
                data[0] * data[7] * data[13] -
                data[4] * data[1] * data[15] +
                data[4] * data[3] * data[13] +
                data[12] * data[1] * data[7] -
                data[12] * data[3] * data[5];
        temp.data[14] = -data[0] * data[5] * data[14] +
                data[0] * data[6] * data[13] +
                data[4] * data[1] * data[14] -
                data[4] * data[2] * data[13] -
                data[12] * data[1] * data[6] +
                data[12] * data[2] * data[5];
        temp.data[3] = -data[1] * data[6] * data[11] +
                data[1] * data[7] * data[10] +
                data[5] * data[2] * data[11] -
                data[5] * data[3] * data[10] -
                data[9] * data[2] * data[7] +
                data[9] * data[3] * data[6];
        temp.data[7] = data[0] * data[6] * data[11] -
                data[0] * data[7] * data[10] -
                data[4] * data[2] * data[11] +
                data[4] * data[3] * data[10] +
                data[8] * data[2] * data[7] -
                data[8] * data[3] * data[6];
        temp.data[11] = -data[0] * data[5] * data[11] +
                data[0] * data[7] * data[9] +
                data[4] * data[1] * data[11] -
                data[4] * data[3] * data[9] -
                data[8] * data[1] * data[7] +
                data[8] * data[3] * data[5];
        temp.data[15] = data[0] * data[5] * data[10] -
                data[0] * data[6] * data[9] -
                data[4] * data[1] * data[10] +
                data[4] * data[2] * data[9] +
                data[8] * data[1] * data[6] -
                data[8] * data[2] * data[5];

	    final float det = data[0] * temp.data[0] + data[1] * temp.data[4] +
                data[2] * temp.data[8] + data[3] * temp.data[12];
        return temp.div(det);
    }

    public static Mat4 translation(Float3 translation) {
        Mat4 temp = new Mat4();
        temp.data[ 3] = translation.x;
        temp.data[ 7] = translation.y;
        temp.data[11] = translation.z;
        return temp;
    }

    public static Mat4 rotation(Float3 rotation) {
        final float xSin = (float) Math.sin(Math.toRadians(rotation.x));
        final float xCos = (float) Math.cos(Math.toRadians(rotation.x));
        Mat4 xRot = new Mat4();
        xRot.data[ 5] =  xCos;
        xRot.data[ 6] = -xSin;
        xRot.data[ 9] =  xSin;
        xRot.data[10] =  xCos;

        final float ySin = (float) Math.sin(Math.toRadians(rotation.y));
        final float yCos = (float) Math.cos(Math.toRadians(rotation.y));
        Mat4 yRot = new Mat4();
        yRot.data[ 0] =  yCos;
        yRot.data[ 2] =  ySin;
        yRot.data[ 8] = -ySin;
        yRot.data[10] =  yCos;

        final float zSin = (float) Math.sin(Math.toRadians(rotation.z));
        final float zCos = (float) Math.cos(Math.toRadians(rotation.z));
        Mat4 zRot = new Mat4();
        zRot.data[0] =  zCos;
        zRot.data[1] = -zSin;
        zRot.data[4] =  zSin;
        zRot.data[5] =  zCos;

        return zRot.multiply(yRot).multiply(xRot);
    }

    public static Mat4 scaling(Float3 size) {
        Mat4 temp = new Mat4();
        temp.data[ 0] = size.x;
        temp.data[ 5] = size.y;
        temp.data[10] = size.z;
        return temp;
    }

    private static final float halfAngleRad = (float)(Math.PI / 360);
    public static Mat4 perspective(float fov, float ar, float zNear, float zFar) {
		final float tanHalf = 1.0f / (float) Math.tan(fov * halfAngleRad);
        Mat4 temp = new Mat4();
        temp.data[ 0] = tanHalf / ar;
        temp.data[ 5] = tanHalf;
        temp.data[10] = (-zFar - zNear) / (zNear - zFar);
        temp.data[11] = (2 * zNear * zFar) / (zNear - zFar);
        temp.data[14] = 1;
        temp.data[15] = 0;
        return temp;
    }

    public static Mat4 orthographic(float left, float right, float bottom, float top, float nearZ, float farZ) {
        Mat4 temp = new Mat4();
        temp.data[ 0] = 2 / (right - left);
        temp.data[ 5] = 2 / (top - bottom);
        temp.data[10] = -2 / (farZ - nearZ);
        temp.data[ 3] = -(right + left) / (right - left);
        temp.data[ 7] = -(top + bottom) / (top - bottom);
        temp.data[11] = -(farZ + nearZ) / (farZ - nearZ);
        return temp;
    }

    public static Mat4 lookAt(Float3 pos, Float3 tar, Float3 up) {
        final Float3 f = tar.subtract(pos).normalize();
        final Float3 s = up.cross(f).normalize();
        final Float3 u = f.cross(s);
        Mat4 temp = new Mat4();
        temp.data[ 0] = s.x;
        temp.data[ 1] = s.y;
        temp.data[ 2] = s.z;
        temp.data[ 3] = -s.dot(pos);
        temp.data[ 4] = u.x;
        temp.data[ 5] = u.y;
        temp.data[ 6] = u.z;
        temp.data[ 7] = -u.dot(pos);
        temp.data[ 8] = f.x;
        temp.data[ 9] = f.y;
        temp.data[10] = f.z;
        temp.data[11] = -f.dot(pos);
        return temp;
    }

    public String toString() {
        return  "[" + data[ 0] + " " + data[ 1] + " " + data[ 2] + " " + data[ 3] + "]\n" +
                "[" + data[ 4] + " " + data[ 5] + " " + data[ 6] + " " + data[ 7] + "]\n" +
                "[" + data[ 8] + " " + data[ 9] + " " + data[10] + " " + data[11] + "]\n" +
                "[" + data[12] + " " + data[13] + " " + data[14] + " " + data[15] + "]";
    }
}
